lib/experimental/new_embed.dart

```diff
@@ -245,81 +268,110 @@ class NewEmbed {
-      final request = dartServices.analyze(input).timeout(serviceCallTimeout);
+      final userSource = context.dartSource;
+      // Create a synthesis of the user code and other code to analyze.
+      final fullSource = '$userSource\n'
+          '${context.testMethod}\n'
+          '${executionSvc.testResultDecoration}\n'
+          'var resultFunction = _result;\n';
+      final sourceRequest = SourceRequest()..source = fullSource;
+      final lines = Lines(sourceRequest.source);
 
-      request.then((AnalysisResults result) {
+      dartServices
+          .analyze(sourceRequest)
+          .timeout(serviceCallTimeout)
+          .then((AnalysisResults result) {
         // Discard if the document has been mutated since we requested analysis.
-        if (input.source != userCodeEditor.document.value) return false;
+        if (userSource != context.dartSource) return;
 
         _displayIssues(result.issues);
 
-        userCodeEditor.document
-            .setAnnotations(result.issues.map((AnalysisIssue issue) {
-          final startLine = lines.getLineForOffset(issue.charStart);
-          final endLine =
-              lines.getLineForOffset(issue.charStart + issue.charLength);
+        Iterable<Annotation> issues = result.issues.map((AnalysisIssue issue) {
+          final charStart = issue.charStart;
+          final startLine = lines.getLineForOffset(charStart);
+          final endLine = lines.getLineForOffset(charStart + issue.charLength);
 
           return Annotation(
             issue.kind,
             issue.message,
             issue.line,
             start: Position(
               startLine,
-              issue.charStart - lines.offsetForLine(startLine),
+              charStart - lines.offsetForLine(startLine),
             ),
             end: Position(
               endLine,
-              issue.charStart +
-                  issue.charLength -
-                  lines.offsetForLine(startLine),
+              charStart + issue.charLength - lines.offsetForLine(startLine),
             ),
           );
-        }).toList());
+        });
+
+        userCodeEditor.document.setAnnotations(issues.toList());
+      }).catchError((e) {
+        if (e is! TimeoutException) {
+          final String message = e is ApiRequestError ? e.message : '$e';
+
+          _displayIssues([
+            AnalysisIssue()
+              ..kind = 'error'
+              ..line = 1
+              ..message = message
+          ]);
+          userCodeEditor.document.setAnnotations([]);
+        }
       });
     });
   }
 
+  void _performFormat() async {
+    String originalSource = userCodeEditor.document.value;
+    SourceRequest input = SourceRequest()..source = originalSource;
+
+    try {
+      formatButton.disabled = true;
+      FormatResponse result =
+          await dartServices.format(input).timeout(serviceCallTimeout);
+
+      formatButton.disabled = false;
+
+      if (originalSource != result.newString) {
```

Comment:
Should we abort if the user has changed their source code since hitting the format button?

Reply:
Yup, good point - will update this.